pulp_ansible/app/galaxy/v1/tasks.py

```diff
@@ -0,0 +1,216 @@
+            "github_user": github_user,
+            "github_repo": github_repo,
+            "role_name": role_name,
+            "name": role_name,
+            "namespace": github_user,
+            "role_version": github_reference,
+        }
+
+        # This is a pulp'ism for the source url
+        # but we will generate the download_url in the
+        # serializer so this wouldn't ever be
+        # used ... ?
+        relative_path = "%s/%s/%s.tar.gz" % (
+            metadata["github_user"],
+            metadata["role_name"],
+            metadata["role_version"],
+        )
+
+        # Make a tarball
+        shutil.rmtree(os.path.join(checkout_path, ".git"))
+        tarfn = f"/tmp/{github_user}.{role_name}.{github_reference}.tar.gz"
+        cmd = f"tar czvf {tarfn} {checkout_path}"
+        pid = subprocess.run(cmd, shell=True)
+        assert pid.returncode == 0
+
+        """
+        # pulp_ansible's Roles are actuallly Role Versions
+        # The roles list viewset aggregates the versions together
+        # by using a {namespace}.{name} as the primary key.
+        role = Role(
+            version=metadata["role_version"],
+            name=metadata["name"],
+            namespace=metadata["namespace"],
+        )
+        role.save()
+
+        # Create a real artifact
+        artifact = Artifact.init_and_validate(tarfn)
+        artifact.save()
+
+        # Use Content to tie the role and artifact together
+        ca1 = ContentArtifact.objects.create(
+            artifact=artifact, content=role, relative_path=relative_path
+        )
+        ca1.save()
+
+        # Add the role to the legacy repository via a new version.
+        legacy = AnsibleRepository.objects.get(pulp_id=ansible_repo_id)
+        add_and_remove(legacy.pk, [role.pk], [])
+        """
+
+        _create_role_and_add_to_repo_with_transaction(
+            metadata, relative_path, tarfn, ansible_repo_id
+        )
+
+
+@transaction.atomic
+def _create_role_and_add_to_repo_with_transaction(
+    role_metadata, relative_path, tarfn, ansible_repo_id
+):
+    """
+    Encapsulate database writes with a transaction.
+    """
+    # pulp_ansible's Roles are actuallly Role Versions
+    # The roles list viewset aggregates the versions together
+    # by using a {namespace}.{name} as the primary key.
+    role = Role(
+        version=role_metadata["role_version"],
+        name=role_metadata["name"],
+        namespace=role_metadata["namespace"],
+    )
+    role.save()
+
+    # Create a real artifact
+    artifact = Artifact.init_and_validate(tarfn)
+    artifact.save()
+
+    # Use Content to tie the role and artifact together
+    ca1 = ContentArtifact.objects.create(
+        artifact=artifact, content=role, relative_path=relative_path
+    )
+    ca1.save()
+
+    # Add the role to the legacy repository via a new version.
+    legacy = AnsibleRepository.objects.get(pulp_id=ansible_repo_id)
+    add_and_remove(legacy.pk, [role.pk], [])
```

Comment:
L#215 and L#216 should not be in this transaction I believe.

Reply:
fixed